persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java

```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+    }
+
+    if (StorageService.instance.isRpcReady(endpoint)) {
+      notifyJoinCluster(endpoint);
+    } else {
+      endpointsPendingJoinedNotification.add(endpoint);
+    }
+
+    return true;
+  }
+
+  private void leaveCluster(InetAddress endpoint) {
     if (!liveStargateNodes.remove(endpoint)) {
       return;
     }
+    StargateSystemKeyspace.removeEndpoint(endpoint);
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onLeaveCluster(nativeAddress, EventListener.NO_PORT);
+    }
+  }
 
-    for (IEndpointLifecycleSubscriber subscriber : subscribers) subscriber.onLeaveCluster(endpoint);
+  private void notifyJoinCluster(InetAddress endpoint) {
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onJoinCluster(nativeAddress, EventListener.NO_PORT);
+    }
   }
 
-  @Override
-  public void onRestart(InetAddress endpoint, EndpointState state) {}
+  private void notifyRpcChange(InetAddress endpoint, boolean isReady) {
+    if (isReady) {
+      notifyUp(endpoint);
+    } else {
+      notifyDown(endpoint);
+    }
+  }
 
-  private void addStargateNode(InetAddress endpoint, EndpointState epState) {
-    if (!isStargateNode(epState) || !liveStargateNodes.add(endpoint)) {
+  private void notifyUp(InetAddress endpoint) {
+    if (!StorageService.instance.isRpcReady(endpoint) || !Gossiper.instance.isAlive(endpoint)) {
       return;
     }
 
-    for (IEndpointLifecycleSubscriber subscriber : subscribers) subscriber.onJoinCluster(endpoint);
+    if (endpointsPendingJoinedNotification.remove(endpoint)) {
+      notifyJoinCluster(endpoint);
+    }
+
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onUp(nativeAddress, EventListener.NO_PORT);
+    }
+  }
+
+  private void applyState(
+      InetAddress endpoint, ApplicationState state, VersionedValue value, EndpointState epState) {
+    final ExecutorService executor = StageManager.getStage(Stage.MUTATION);
+    switch (state) {
+      case RELEASE_VERSION:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "release_version", value.value, executor);
+        break;
+      case DC:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "data_center", value.value, executor);
+        break;
+      case RACK:
+        StargateSystemKeyspace.updatePeerInfo(endpoint, "rack", value.value, executor);
+        break;
+      case RPC_ADDRESS:
+        try {
+          StargateSystemKeyspace.updatePeerInfo(
+              endpoint, "rpc_address", InetAddress.getByName(value.value), executor);
+        } catch (UnknownHostException e) {
+          throw new RuntimeException(e);
+        }
+        break;
+      case SCHEMA:
+        // Use a fix schema version for all peers (always in agreement) because stargate waits
```

Comment:
do you mean that there is happens-before relation between those events?

Reply:
Not sure what you mean here.

Reply:
I mean that stargate waits here, but it is related to this https://github.com/stargate/stargate/pull/267#discussion_r510768342 comment, so please disregard this comment